### PR TITLE
Checking the validity of the message id

### PR DESCRIPTION
**Pull request by Tesa27**
When a message is received by the backend the message id was not checked, which created a potential integrity violation. A user could send a message containing the wrong message id or a message that has the message id of a valid message and backend will accept them both if the remaining fields are correctly set.

### CODE REVIEW

File: be1-go/hub/standard_hub/message_handling.go [hunk 1]
```diff
@@ -269,6 +269,17 @@ func (h *Hub) handlePublish(socket socket.Socket, byteMessage []byte) (int, erro
...
 
+	signature := publish.Params.Message.Signature
+	messageId := publish.Params.Message.MessageID
+	data := publish.Params.Message.Data
+
+	expectedMessageId := messagedata.Hash(data, signature)
+
+	if expectedMessageId != messageId {
+
+		return publish.ID, xerrors.Errorf("message_id is wrong expected is %q but was %v", expectedMessageId, messageId)
```
Comment (Warhedgehog) [inline be1-go/hub/standard_hub/message_handling.go:280]: I may be missing something but would it not be massively vulnerable to inform the sender of the hash that we expected ? Could he not resend the same message with the appropriate message_id thus impersonating someone else ?

**Review by Warhedgehog — CHANGES_REQUESTED** [review]
I may misunderstand the message_id but there seems to be a big problem with the current proposed changes.

Nevertheless the check is crucial, so nice change.  I think you should also do it for handleBroadcast below

File: be1-go/hub/standard_hub/message_handling.go (hunk 1 above)
Comment (Tesa27) [reply be1-go/hub/standard_hub/message_handling.go:280]: I do agree, the attacker could just send a random message id to get information about the true message id and would just resend the same message with the correct message id.

**Review by Warhedgehog — COMMENTED** [review]
It would be good to do a unit test for the handleBroadcast too

**Review by Warhedgehog — DISMISSED** [review]
LGTM

File: be1-go/hub/standard_hub/mod_test.go [hunk 2]
```diff
@@ -121,6 +125,91 @@ func Test_Create_LAO_Bad_Key(t *testing.T) {
...
+
+	hub.handleMessageFromClient(&socket.IncomingMessage{
+		Socket:  sock,
+		Message: publishBuf,
+	})
+
+	require.EqualError(
+		t, sock.err,
+		fmt.Sprintf(
+			"failed to handle method: message_id %vis wrong", badMessageID))
```
Comment (Robingoumaz) [inline be1-go/hub/standard_hub/mod_test.go:210]: there is a missing space

**Review by MaximeZmt — APPROVED** [review]
👍

**Review by Robingoumaz — CHANGES_REQUESTED** [review]
Great work, just a little typo

File: be1-go/hub/standard_hub/message_handling.go [hunk 3]
```diff
@@ -269,6 +269,16 @@ func (h *Hub) handlePublish(socket socket.Socket, byteMessage []byte) (int, erro
 		return -1, xerrors.Errorf("failed to unmarshal publish message: %v", err)
 	}
 
+	signature := publish.Params.Message.Signature
+	messageId := publish.Params.Message.MessageID
+	data := publish.Params.Message.Data
+
+	expectedMessageId := messagedata.Hash(data, signature)
```
Comment (lhmerino) [inline be1-go/hub/standard_hub/message_handling.go:276]: From a security perspective, there should be a delimiter between data and signature. https://crypto.stackexchange.com/questions/55162/best-way-to-hash-two-values-into-one

File: be1-go/hub/standard_hub/message_handling.go [hunk 4]
```diff
@@ -269,6 +269,16 @@ func (h *Hub) handlePublish(socket socket.Socket, byteMessage []byte) (int, erro
...
+
+	if expectedMessageId != messageId {
+		return publish.ID, xerrors.Errorf("message_id  is wrong")
```
Comment (lhmerino) [inline be1-go/hub/standard_hub/message_handling.go:279]: Extra space.

File: be1-go/hub/standard_hub/message_handling.go (hunk 3 above)
Comment (lhmerino) [reply be1-go/hub/standard_hub/message_handling.go:276]: Although this change would likely require the cooperation of all subsystems but generally its essential to add a delimiter when concatenating strings that is to be hashed.

**Review by lhmerino — CHANGES_REQUESTED** [review]
LGTM; left some comments.

File: be1-go/hub/standard_hub/message_handling.go (hunk 3 above)
Comment (Tesa27) [reply be1-go/hub/standard_hub/message_handling.go:276]: I see I see, for the moment the delimiter is not included in any subsystem if I am not mistaken, so this should be a global change in the system and I guess in the specifications as well

Comment (pierluca) [reply be1-go/hub/standard_hub/message_handling.go:276]: Actually, we don't need a delimiter because we prefix each data element with its length.

**Review by Robingoumaz — DISMISSED** [review]

File: be1-go/hub/standard_hub/message_handling.go [hunk 5]
```diff
@@ -123,6 +123,7 @@ func (h *Hub) handleRootChannelBroadcastMesssage(sock socket.Socket, broadcast m
 	if err != nil {
 		h.log.Err(err).Msg("invalid lao#create message")
 		sock.SendError(&id, err)
+		return err
```
Comment (nkcr) [inline be1-go/hub/standard_hub/message_handling.go:126]: nice catch! 🙀

File: be1-go/hub/standard_hub/message_handling.go [hunk 6]
```diff
@@ -269,6 +270,15 @@ func (h *Hub) handlePublish(socket socket.Socket, byteMessage []byte) (int, erro
 		return -1, xerrors.Errorf("failed to unmarshal publish message: %v", err)
 	}
 
+	signature := publish.Params.Message.Signature
+	messageId := publish.Params.Message.MessageID
+	data := publish.Params.Message.Data
+
+	expectedMessageId := messagedata.Hash(data, signature)
+	if expectedMessageId != messageId {
+		return publish.ID, xerrors.Errorf("message_id is wrong")
```
Comment (nkcr) [inline be1-go/hub/standard_hub/message_handling.go:279]:
```suggestion
		return publish.ID, xerrors.Errorf("message_id is wrong: expected %q found %q", expectedMessageId, messageId)
```

Comment (nkcr) [reply be1-go/hub/standard_hub/message_handling.go:279]: same bellow

File: be1-go/hub/standard_hub/mod_test.go [hunk 7]
```diff
@@ -677,6 +779,90 @@ func Test_Create_LAO_Broadcast(t *testing.T) {
...
+	keypair := generateKeyPair(t)
+
+	fakeChannelFac := &fakeChannelFac{
+		c: &fakeChannel{},
+	}
+
+	hub, err := NewHub(keypair.public, nolog, fakeChannelFac.newChannel, hub.OrganizerHubType)
+	require.NoError(t, err)
+
+	now := time.Now().Unix()
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod_test.go:793]: tests should be reproducible. Can't we use a fixed value ?

File: be1-go/hub/standard_hub/mod_test.go [hunk 8]
```diff
@@ -677,6 +779,90 @@ func Test_Create_LAO_Broadcast(t *testing.T) {
...
+	require.NoError(t, err)
+
+	sock := &fakeSocket{}
+
+	hub.handleMessageFromServer(&socket.IncomingMessage{
+		Socket:  sock,
+		Message: publishBuf,
+	})
+
+	require.EqualError(
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod_test.go:860]: Mmmm we avoid that type of formatting. 
In the case of a test it is fine to have
```go
require.EqualError(t, sock.err, fmt.Sprintf("failed to handle method: message_id %v is wrong", fakeMessageID))
```

Comment (nkcr) [reply be1-go/hub/standard_hub/mod_test.go:860]: If that wasn't a test, we would format it as
```go
	require.EqualError(t, sock.err, fmt.Sprintf("failed to handle method: "+
		"message_id %v is wrong", fakeMessageID))
```

**Review by nkcr — COMMENTED** [review]
Good job, have a couple of minor comments 👍

**Review by nkcr — APPROVED** [review]

**Pull request merged**